### PR TITLE
Fix compatibility with doctrine/dbal 2.13

### LINKED ISSUE #2928: Authentication request failed
**Affected version(s)**

Contao 4.11

**Description**

Wenn man mittels Contao-Console einen neuen Nutzer anlegt:
```bash
vendor/bin/contao-console contao:user:create --username={{ item[0] }} --name='{{ item[1] }}' --email='{{ item[2] }}' --password='{{ item[3] }}' --change-password {{ item[5] }} --language={{ item[4] }}
```
Dann kann dieser sich nicht anmelden, auch der Counter wird nicht hochgezählt.

Im Log steht folgendes
```
[2021-03-30T14:36:08.218624+02:00] request.INFO: Matched route "contao_backend_login". {"route":"contao_backend_login","route_parameters":{"_route":"contao_backend_login","_scope":"backend","_token_check":true,"_controller":"Contao\\CoreBundle\\Controller\\BackendController::loginAction"},"request_uri":"/contao/login?_hash=LDH7EL6CQiJ95VKcE%2BLDEWJZOACQ3XMqWm8Kzw8JaQ8%3D&redirect=%2Fcontao%2F","method":"POST"} []
[2021-03-30T14:36:08.230412+02:00] security.INFO: Authentication request failed. {"exception":"[object] (Symfony\\Component\\Security\\Core\\Exception\\AuthenticationServiceException(code: 0): $statement must be a Statement object or an array at /var/customers/webs/vendor/symfony/security-core/Authentication/Provider/DaoAuthenticationProvider.php:96)\n[previous exception] [object] (InvalidArgumentException(code: 0): $statement must be a Statement object or an array at /var/customers/webs/vendor/contao/core-bundle/src/Resources/contao/library/Contao/Database/Result.php:93)"} []
[2021-03-30T14:36:08.262293+02:00] request.CRITICAL: Uncaught PHP Exception InvalidArgumentException: "$statement must be a Statement object or an array" at /var/customers/webs/vendor/contao/core-bundle/src/Resources/contao/library/Contao/Database/Result.php line 93 {"exception":"[object] (InvalidArgumentException(code: 0): $statement must be a Statement object or an array at /var/customers/webs/vendor/contao/core-bundle/src/Resources/contao/library/Contao/Database/Result.php:93)"} []
```

### PR DESCRIPTION
**Pull request by ausi**
| Q                | A
| -----------------| ---
| Fixed issues     | Fixes #2928

I’m not yet sure if this is the correct fix, or a doctrine bug, or both.

### CODE REVIEW

**Comment by leofeyer** [comment]
At least it seems to fix the issue. The functional tests are failing in the `4.9` branch now with doctrine/dbal in version 2.13.0 (see https://github.com/contao/contao/pull/2918/checks?check_run_id=2229086818) but they do not fail here.

**Comment by ausi** [comment]
I think this fix is correct now. 

`doctrine/dbal` in version `2.8.0` changed the return type of `Doctrine\DBAL\Connection::executeQuery()` from `Doctrine\DBAL\Driver\Statement` to `Doctrine\DBAL\Driver\ResultStatement`, see https://github.com/doctrine/dbal/commit/df4343721d9981e58afd15eeaf18c885a2399766#diff-425f50e688c77962b97e5fd1f9803a6d9ffa9bf260ba12391f73b2aab14914eeL906-R911

This pull request now updates the type hints accordingly and adds a type check before calling `->rowCount()` to make sure this method is actually present.

File: core-bundle/src/Resources/contao/library/Contao/Database/Result.php [hunk 1]
```diff
@@ -363,7 +366,13 @@ public function count()
 	{
 		if ($this->rowCount === null)
 		{
-			$this->rowCount = $this->resResult->rowCount();
+			if (
+				$this->resResult instanceof DoctrineStatement
+				|| $this->resResult instanceof DoctrineResult
+				|| $this->resResult instanceof DoctrineDriverResult
```
Comment (leofeyer) [inline core-bundle/src/Resources/contao/library/Contao/Database/Result.php:372]: `Doctrine\DBAL\Driver\Result` does not extend `Doctrine\DBAL\Driver\ResultStatement`, so based on the if statement in line 85, `$this->resResult instanceof DoctrineDriverResult ` can never be `true`, can it?

Comment (leofeyer) [inline core-bundle/src/Resources/contao/library/Contao/Database/Result.php]: `Doctrine\DBAL\Driver\Result` does not extend `Doctrine\DBAL\Driver\ResultStatement`, so based on the if statement in line 85, this condition can never be `true`, can it?

File: core-bundle/src/Resources/contao/library/Contao/Database/Result.php [hunk 2]
```diff
@@ -363,7 +366,13 @@ public function count()
 	{
 		if ($this->rowCount === null)
 		{
-			$this->rowCount = $this->resResult->rowCount();
+			if (
+				$this->resResult instanceof DoctrineStatement
+				|| $this->resResult instanceof DoctrineResult
```
Comment (leofeyer) [inline core-bundle/src/Resources/contao/library/Contao/Database/Result.php]: `Doctrine\DBAL\Result` only exists in doctrine/dbal 3.0, so I assume this is a forward-compatibility layer, right? Still I wonder how this condition can ever be `true` given the if statement in line 85.

File: core-bundle/src/Resources/contao/library/Contao/Database/Result.php (hunk 1 above)
Comment (ausi) [reply core-bundle/src/Resources/contao/library/Contao/Database/Result.php]: `Doctrine\DBAL\Driver\Result` is an interface, so there can be objects that implement both `Result` and `ResultStatement`.

File: core-bundle/src/Resources/contao/library/Contao/Database/Result.php (hunk 2 above)
Comment (ausi) [reply core-bundle/src/Resources/contao/library/Contao/Database/Result.php]: Exists in 2.13.0 too https://github.com/doctrine/dbal/blob/2.13.x/lib/Doctrine/DBAL/Result.php

Comment (leofeyer) [reply core-bundle/src/Resources/contao/library/Contao/Database/Result.php]: Ah, new file structure in version 3. 👍🏻

File: core-bundle/src/Resources/contao/library/Contao/Database/Result.php (hunk 1 above)
Comment (ausi) [reply core-bundle/src/Resources/contao/library/Contao/Database/Result.php]: Changed to a `method_exists` check in 6d4ab06db2f96877e30b44749b164c90eeb4e39b

**Review by leofeyer — DISMISSED** [review]

**Comment by leofeyer** [comment]
Thank you @ausi.

**Pull request merged**